keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/PersistenceExceptions.kt

```diff
@@ -0,0 +1,54 @@
+package com.netflix.spinnaker.keel.exceptions
+
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.kork.exceptions.SystemException
+import com.netflix.spinnaker.kork.exceptions.UserException
+
+sealed class NoSuchEntityException(override val message: String?) :
```

Comment:
I've consolidated all exceptions from the various repositories in this file so that it's easy to follow the hierarchy, and have refactored the hierarchy so that all "not found" type exceptions inherit from `NoSuchEntityException`. This allows me to handle the "not found" case in a generic way in `AuthorizationSupport` so that we can allow 404s to be returned from the APIs instead of 403s. To be honest, I'm not sure that's the right thing to do -- if you don't have permission, in theory we shouldn't even query the database at all. Thoughts?

Reply:
If the resource in question doesn't exist - a delivery config, and application, etc - we should return a `..NotFound` instead of failing because you "aren't authorized". If the thing does exist enough to check permissions and you're not authorized, we should return unauthorized

Reply:
I'm generally not a fan of packaging things according to what type of thing they are (exceptions in this case) rather than what bit of functionality they're involved in. Is there any reason to not co-locate these with the repository interfaces themselves?

Reply:
Okey dokey! So this is as should be. 😃

Reply:
@robfletcher Not technically. I just thought it would be very easy to forget to extend `NoSuchEntityException` if it's not near the other exception types. But I agree with you on the principle for code organization in general. Let me see how to better organize them.

P.S. In my defense, we already had a `com.netflix.spinnaker.keel.exceptions` package elsewhere.